File: src/engine.js
import * as lookups from './lookups.js';
import { bigIntContainerToObject } from './gramutil.js';

class Matcher {
    constructor (board, x, y, dir) {
        this.board = board;
        this.x = x;
        this.y = y;
        this.dir = lookups.charLookup.absDir[dir];
        this.termAddr = [];
        this.termCell = [];
        this.termTailStart = [];
        this.failed = false;
    }

    getCell (x, y) { return board.getCell (x + this.x, y + this.y); }

    matchLhsTerm (t, type, state) {
        if (t.op === 'any')
            return true;
        if (t.op === 'negterm')
            return !this.matchLhsTerm (t.term, type, state);
        if (t.op === 'alt')
            return t.alt.reduce ((matched, term) => matched || this.matchLhsTerm (term, type, state), false);
        if (t.type !== type)
            return false;

        if (!t.state)
          return !state.length;

        for (let n = 0; n < t.state.length; ++n) {
          const matchStatus = this.matchStateChar (t.state[n], state.charAt(n));
          if (!matchStatus)
            return false;
          if (matchStatus < 0)
            return true;
        }
        return t.state.length === state.length;
    }

    // return true to match char, -1 to match all remaining state chars
    matchStateChar (s, c) {
      if (typeof(s) === 'string')
        return s === c;
      switch (s.op) {
      case 'char':
        return s.char === c;
      case 'wild':
        return typeof(c) !== 'undefined';
      case 'any':
        return -1;  // match all remaining state chars
      case 'class':
        return s.chars.indexOf(c) >= 0;
      case 'negated':
        return s.chars.indexOf(c) < 0;
      default:
        return this.computeStateChar (s) === c;
      }
    }

    computeStateChar (t) {
      if (typeof(t) === 'string')
        return t;
      switch (t.op) {
      case 'char':
        return t;
      case 'clock':
      case 'anti':
        return lookups.charPermLookup.rotate[t.op][this.computeStateChar(t.arg)];
      case 'add':
        return lookups.charPermLookup.intAdd[this.computeStateChar(t.right)][this.computeStateChar(t.left)];
      case 'sub':
      case '+':
        return lookups.charPermLookup.vecAdd[this.computeStateChar(t.right)][this.computeStateChar(t.left)];
      case '-':
        return lookups.charPermLookup.vecSub[this.computeStateChar(t.right)][this.computeStateChar(t.left)];
      case '*':
        return lookups.charPermLookup.matMul[t.left.matrix][this.computeStateChar(t.right)];
      case 'location':
        return lookups.vec2char (this.termAddr[t.group-1]);
      case 'reldir':
        return this.getRelativeDir (t.dir);
      case 'absdir':
        return lookups.charVecLookup[t.dir];
      case 'integer':
        return lookups.int2char (t.n);
      case 'vector':
        return lookups.vec2char (t.x, t.y);
      case 'state':
        return this.termCell[t.group-1].state.charAt(t.char-1);
      case 'tail':
        return this.termCell[t.group-1].state.substr(this.termTailStart[t.group-1]);
      default:
        throw new Error ("Unrecognized op '" + t.op + "' in " + JSON.stringify(t));
      }
    }

    getRelativeDir (dir) {
        return lookups.charPermLookup.matMul[dir][this.dir];
    }

    computeAddr (addr, baseVec) {
      switch (addr.op) {
        case 'absolute':
            return lookups.charVecLookup[lookups.charPermLookup.vecAdd[lookups.absDir[t.dir]][lookups.vec2char(baseVec)]];
        case 'relative':
            return lookups.charVecLookup[lookups.charPermLookup.vecAdd[this.getRelativeDir(addr.dir)][lookups.vec2char(baseVec)]];
        case 'neighbor':
            return lookups.charVecLookup[lookups.charPermLookup.vecAdd[this.computeStateChar(addr.arg)][lookups.vec2char(baseVec)]];
        case 'cell':
            return lookups.charVecLookup[this.computeStateChar(addr.arg)];
        default:
          throw new Error ("Unrecognized op '" + addr.op + "' in " + JSON.stringify(addr));
      }
    }

    matchLhsCell (term, pos) {
        if (!this.failed) {
            let x, y;
            if (pos === 0)
                x = y = 0;
            else
                [x,y] = this.computeAddr (term.dir || { op: 'relative', dir: 'F' }, this.termAddr[pos-1], pos);
            this.termAddr.push ([x,y]);
            const cell = this.board.getCell (x + this.x, y + this.y);
            const { type, state } = cell;
            const match = this.matchLhsTerm (term, type, state);
            if (match) {
                this.termCell.push (cell);
                this.termTailStart.push (term.state && term.state[term.state.length-1].op === 'any' ? term.state.length - 1 : state.length);
            } else
                this.failed = true;
        } else
            this.failed = true;
        return this;
    }

    getLhsPosForRhsTerm (t) {
      if (t.id)
        return t.id;
      if (t.op === 'group' || t.op === 'prefix')
        return t.group;
      return undefined;
    }

    getMetaForRhsTerm (t, score) {
      const g = this.getLhsPosForRhsTerm (t);
      if (g) {
        const meta = this.termCell[g-1].meta;
        if (meta || score)
          return { ...(meta||{}), ...(g===1 && score ? { score: (meta?.score || 0) + score } : {}) };
      }
      return undefined;
    }

    newCell (t, score) {
      const meta = this.getMetaForRhsTerm (t, score);
      if (t.op === 'group') {
            const { type, state } = this.termCell[t.group-1];
            return { type, state, meta }
        }
        if (t.op === 'prefix') {
            const { type } = this.termCell[t.group-1];
            const state = t.state ? t.state.map(this.computeStateChar.bind(this)).join('') : '';
            return { type, state, meta }
        }
        const { type } = t;
        const state = t.state ? t.state.map(this.computeStateChar.bind(this)).join('') : '';
        return { type: t.type, state, meta };
    }
    
    newCellUpdate (term, pos, score) {
      const a = this.termAddr[pos];
      return [a[0] + this.x, a[1] + this.y, this.newCell (term, score)];
    }
};

const applyTransformRule = (board, x, y, dir, rule) => {
  const updates = transformRuleUpdate (board, x, y, dir, rule);
  if (updates)
    updates.forEach ((update) => update && board.setCell (...update))
  return !!updates;
}

const stripDuplicateMetadata = (matcher, domCell, subCell, domTerm, subTerm) => {
  if (matcher.getLhsPosForRhsTerm(domTerm) === matcher.getLhsPosForRhsTerm(subTerm))
    delete subCell.meta;
  if (subCell?.meta?.id && domCell?.meta?.id === subCell?.meta?.id)
    delete subCell.meta['id'];
  if (subCell.meta && !Object.keys(subCell.meta).length)
    delete subCell.meta;
}

const matchLhs = (board, x, y, dir, rule) => rule.lhs.reduce ((matcher, term, pos) => matcher.matchLhsCell(term,pos), new Matcher (board, x, y, dir) );
const transformRuleUpdate = (board, x, y, dir, rule) => {
  const matcher = matchLhs (board, x, y, dir, rule);
  if (matcher.failed) return null;
  let update = rule.rhs.map ((term, pos) => matcher.newCellUpdate(term,pos,rule.score));
  for (let i = 0; i < update.length-1; ++i)
    for (let j = i + 1; j < update.length; ++j)
      stripDuplicateMetadata (matcher, update[i][2], update[j][2], rule.rhs[i], rule.rhs[j]);
  return update;
}

export { applyTransformRule, transformRuleUpdate, matchLhs };

File: python/engine.py
import lookups
from gramutil import bigIntContainerToObject

class Matcher:
    def __init__(self, board, x, y, dir):
        self.board = board
        self.x = x
        self.y = y
        self.dir = lookups.charLookup.absDir[dir]
        self.termAddr = []
        self.termCell = []
        self.termTailStart = []
        self.failed = False

    def getCell(self, x, y):
        return self.board.getCell(x + self.x, y + self.y)

    def matchLhsTerm(self, t, type, state):
        if t['op'] == 'negterm':
            return not self.matchLhsTerm(t['term'], type, state)
        if t['op'] == 'alt':
            return any(self.matchLhsTerm(term, type, state) for term in t['alt'])
        if t['type'] != type:
            return False
        if not t['state']:
            return not state
        for n in range(len(t['state'])):
            matchStatus = self.matchStateChar(t['state'][n], state[n])
            if not matchStatus:
                return False
            if matchStatus < 0:
                return True
        return len(t['state']) == len(state)

    def matchStateChar(self, s, c):
        if isinstance(s, str):
            return s == c
        op = s['op']
        if op == 'char':
            return s['char'] == c
        if op == 'wild':
            return isinstance(c, str)
        if op == 'any':
            return -1
        if op == 'class':
            return c in s['chars']
        if op == 'negated':
            return c not in s['chars']
        return self.computeStateChar(s) == c

    def computeStateChar(self, t):
        if isinstance(t, str):
            return t
        op = t['op']
        if op == 'char':
            return t
        if op == 'clock' or op == 'anti':
            return lookups.charPermLookup.rotate[op][self.computeStateChar(t['arg'])]
        if op == 'add':
            return lookups.charPermLookup.intAdd[self.computeStateChar(t['right'])][self.computeStateChar(t['left'])]
        if op == 'sub' or op == '+':
            return lookups.charPermLookup.vecAdd[self.computeStateChar(t['right'])][self.computeStateChar(t['left'])]
        if op == '-':
            return lookups.charPermLookup.vecSub[self.computeStateChar(t['right'])][self.computeStateChar(t['left'])]
        if op == '*':
            return lookups.charPermLookup.matMul[t['left']['matrix']][self.computeStateChar(t['right'])]
        if op == 'location':
            return lookups.vec2char(self.termAddr[t['group']-1])
        if op == 'reldir':
            return self.getRelativeDir(t['dir'])
        if op == 'absdir':
            return lookups.charVecLookup[t['dir']]
        if op == 'integer':
            return lookups.int2char(t['n'])
        if op == 'vector':
            return lookups.vec2char(t['x'], t['y'])
        if op == 'state':
            return self.termCell[t['group']-1]['state'][t['char']-1]
        if op == 'tail':
            return self.termCell[t['group']-1]['state'][self.termTailStart[t['group']-1]:]
        raise ValueError(f"Unrecognized op '{op}' in {t}")

    def getRelativeDir(self, dir):
        return lookups.charPermLookup.matMul[dir][self.dir]

    def computeAddr(self, addr, baseVec):
        op = addr['op']
        if op == 'absolute':
            return lookups.charVecLookup[lookups.charPermLookup.vecAdd[lookups.absDir[self.dir]][lookups.vec2char(baseVec)]]
        if op == 'relative':
            return lookups.charVecLookup[lookups.charPermLookup.vecAdd[self.getRelativeDir(addr['dir'])][lookups.vec2char(baseVec)]]
        if op == 'neighbor':
            return lookups.charVecLookup[lookups.charPermLookup.vecAdd[self.computeStateChar(addr['arg'])][lookups.vec2char(baseVec)]]
        if op == 'cell':
            return lookups.charVecLookup[self.computeStateChar(addr['arg'])]
        raise ValueError(f"Unrecognized op '{op}' in {addr}")

    def matchLhsCell(self, term, pos):
        if not self.failed:
            if pos == 0:
                x = y = 0
            else:
                x, y = self.computeAddr(term.get('dir', {'op': 'relative', 'dir': 'F'}), self.termAddr[pos-1], pos)
            self.termAddr.append([x, y])
            cell = self.board.getCell(x + self.x, y + self.y)
            type = cell['type']
            state = cell['state']
            match = self.matchLhsTerm(term, type, state)
            if match:
                self.termCell.append(cell)
                self.termTailStart.append(len(state) - 1 if term.get('state') and term['state'][-1]['op'] == 'any' else len(state))
            else:
                self.failed = True
        else:
            self.failed = True
        return self

    def newCell(self, t):
        if t['op'] == 'group':
            cell = self.termCell[t['group']-1]
            return {'type': cell['type'], 'state': cell['state'], 'meta': cell['meta']}
        if t['op'] == 'prefix':
            cell = self.termCell[t['group']-1]
            newState = ''.join(self.computeStateChar(s) for s in t.get('state', ''))
            return {'type': cell['type'], 'state': newState, 'meta': cell['meta']}
        state = ''.join(self.computeStateChar(s) for s in t.get('state', ''))
        return {'type': t['type'], 'state': state}

    def newCellUpdate(self, term, pos):
        a = self.termAddr[pos]
        return [a[0] + self.x, a[1] + self.y, self.newCell(term)]

def applyTransformRule(board, x, y, dir, rule):
    updates = transformRuleUpdate(board, x, y, dir, rule)
    if updates:
        for update in updates:
            if update:
                board.setCell(*update)
    return bool(updates)

def stripDuplicateMetadata(domCell, subCell):
    if domCell.get('meta') and subCell.get('meta'):
        for prop in ['id', 'owner']:
            if domCell['meta'].get(prop) == subCell['meta'].get(prop):
                del subCell['meta'][prop]
        if not subCell['meta']:
            del subCell['meta']

def transformRuleUpdate(board, x, y, dir, rule):
    matcher = reduce(lambda matcher, term, pos: matcher.matchLhsCell(term, pos), rule['lhs'], Matcher(board, x, y, dir))
    if matcher.failed:
        return None
    update = [matcher.newCellUpdate(term, pos) for pos, term in enumerate(rule['rhs'])]
    for i in range(len(update)-1):
        for j in range(i + 1, len(update)):
            stripDuplicateMetadata(update[i][2], update[j][2])
    return update

__all__ = ['applyTransformRule', 'transformRuleUpdate']


